### PR TITLE
Offers

### PR DESCRIPTION
**Pull request by rpcrezel**

### CODE REVIEW

File: app/persona/api/offer/Offer.scala [hunk 1]
```diff
@@ -1,16 +1,26 @@
...
-             creationTime: Long,
-             expirationTime: Long,
+             expirationTime: DateTime,
              currentParticipants: Int,
-             maxParticipants: Int) {
+             maxParticipants: Int,
+             value: Double,
+             criteria: Map[String,String]) {
+
+  if(criteria.isEmpty) {
```
Comment (taylorstark) [inline app/persona/api/offer/Offer.scala]: Recommend changing to require(criteria.nonEmpty).  That way we don't need InvalidOfferException

File: app/persona/api/offer/offerImpl/OfferSchema.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+class OfferSchema private (val creationDay: DateTime,
+                            val description: String,
+                            val expirationTime: DateTime,
+                            val currentParticipants: Int,
+                            val maxParticipants: Int,
+                            val value: Double,
+                            criteriaDescriptors: Map[String, OfferCriterionDescriptor]) {
+
+  if(description.isEmpty) {
```
Comment (taylorstark) [inline app/persona/api/offer/offerImpl/OfferSchema.scala]: Recommend changing to using require as per previous example

File: app/persona/api/offer/offerImpl/JsonOfferWriter.scala [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      (JsPath \ "creationDay").write[DateTime] and
+      (JsPath \ "description").write[String] and
+      (JsPath \ "expirationTime").write[DateTime] and
+      (JsPath \ "currentParticipants").write[Int] and
+      (JsPath \ "maxParticipants").write[Int] and
+      (JsPath \ "value").write[Double] and
+      (JsPath \ "criteria").write[Map[String, String]]
+    )(unlift(Offer.unapply))
+
+  def toJson(offer: Offer): JsValue = Json.toJson(offer)
```
Comment (taylorstark) [inline app/persona/api/offer/offerImpl/JsonOfferWriter.scala]: Add in another method called def toJson(offers: Seq[Offer]): JsValue = Json.toJson(offers)

Then in the offers controller, we don't need to do 2 casts to json.  Saves us some memory allocations

File: app/persona/controllers/Offers.scala [hunk 4]
```diff
@@ -3,38 +3,65 @@ package persona.controllers
...
+  def get(id: UUID) = Action.async {
+
+    offerService.get(id).map { optionOffer =>
+      optionOffer map { offer =>
+        val jsonOffers = optionOffer.map(jsonOfferWriter.toJson)
+        val serializedJson = Json.toJson(jsonOffers)
+
+        Ok(serializedJson)
       } getOrElse {
         InternalServerError
```
Comment (taylorstark) [inline app/persona/controllers/Offers.scala]: 404 Not found might be better than InternalServerError.  Or return 200 OK with an error message.  Should do some quick research and talking with @d77kim to determine which approach is more appropriate (for handling by clients)

File: test/persona/api/offers/OfferServiceImplSpec.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      offer.description returns "desc"
+      offer.expirationTime returns new DateTime(1437090963326L)
+      offer.currentParticipants returns 2
+      offer.maxParticipants returns 10
+      offer.value returns 3.5
+
+      val offerDao = mock[OfferDAO]
+      offerDao.list(any[ExecutionContext]) returns Future.successful(Seq(offer))
+
+      val dataSchema = mock[OfferSchema]
```
Comment (taylorstark) [inline test/persona/api/offers/OfferServiceImplSpec.scala]: Is this variable used?

File: test/persona/api/offers/offerImpl/JsonOfferParserSpec.scala [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+          |  "value":3.50,
+          |  "criteria":[
+          |    {
+          |      "criterionCategory":"criterion1Cat",
+          |      "criterion":"crit"
+          |    }]
+          |}
+        """.stripMargin
+
+            val tryDataSchema = new JsonOfferParser().parse(json)
```
Comment (taylorstark) [inline test/persona/api/offers/offerImpl/JsonOfferParserSpec.scala]: Minor: Pick better names for variables.  This returns an offer schema, not a data schema

**Comment by taylorstark** [comment]
That's all my comments.  Would be nice if we could get the code coverage up, but if you're busy with other assignments then don't worry about it too much

**Comment by taylorstark** [comment]
2 minor comments.  Don't know if we need InvalidOfferSchema.scala anymore.  In the offers controller, we now return a 200 ok instead of an internal server error.  Should it return json such as:
{
  "status":"error"
  "errors":[
    {
      "errorMessage":"could not find offer xxxxx"
    }
  ]
}

If so, have a look at ControllerHelper.  If we do but you don't have time, create a ticket on pivotal tracker so we can do it later.

Other than that, LGTM

File: app/persona/api/offer/offerImpl/CassandraOfferDataDAO.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+}
+
+
+class CassandraOfferDataDAO extends OfferDataTable with OfferDAO with PersonaCassandraConnector  {
+  // TODO:limit actually 25 for these two?
+  def list(implicit ec: ExecutionContext): Future[Seq[Offer]] = {
+    select
+      .limit(25)
```
Comment (m26jiang) [inline app/persona/api/offer/offerImpl/CassandraOfferDataDAO.scala]: Maybe find a way to pull from config down the road

**Comment by m26jiang** [comment]
looks good otherwise

**Pull request merged**